### DISCUSSION TITLE
esp32 analog input

### DISCUSSION THREAD

**Discussion by micrcx** [toitlang/toit#1036] [Q&A] [answered]
My question is about using the esp32 analog input:

```
import gpio
import gpio.adc

pinSP    := gpio.Pin 36
sensor   := adc.Adc pinSP

main:
  128.repeat :
     print sensor.get
     sleep --ms=1

```
I use it to get data from an ECG sensor (https://www.espressif.com/en/news/ESP32_Heart_Monitoring) and a pulse sensor (https://microcontrollerslab.com/pulse-sensor-esp32-tutorial/). As I understood from the description of ESP32, the analog input that I use allows to receive data as an integer in the range [0...4096] (ADC 12 bits). When reading I get however a set of real numbers. This means that a conversion is performed somewhere and integer values are replaced with real values. Is it possible to get the original values, i.e. rowdata? If obtaining of row data is impossible, then what is the range of real values obtained?

**Comment by floitsch** [comment, marked as answer]
The value returned by the ADC converter is the actual voltage.
It uses calibration data that is stored inside the ESP32 to correct the raw data to be more accurate.
The range of the returned value is fundamentally 0.0 to 3.3, since that's the voltage the ESP32 works with.

I just created a PR to also obtain the raw value of a measurement: https://github.com/toitlang/toit/pull/1037
It's likely going to be included in the next release.

**Reply by micrcx** [reply]
Thanks.

**Reply by floitsch** [reply]
New version has been released.
Use `pin.get --raw` to make one raw measurement.
